### DISCUSSION TITLE
nuxtServerInit and static, rate limiting

### DISCUSSION THREAD

**Discussion by CHEWX** [nuxt/nuxt#10282] [Questions (Nuxt 2)]
### Steps to reproduce

A simple axios get in nuxtServerInit in store.

### What is Expected?

Each generated page should fire once, not re-fire

### What is actually happening?

Re-firing every time.

I've found this, but it's 3+ years old and not sure how relevant it is - https://github.com/nuxt/nuxt.js/issues/3935

Is there a clear way to do this? Cannot find anything that works, I've tried caching but that seems to reset every route generation. A service such as BigCommerce only allows 150 calls every 30 seconds and when you have over 150 pages, 1 single route hits the rate limit as it's being called on every page generation.

Caching method:

~/plugins/axios.js

```
import { cacheAdapterEnhancer } from 'axios-extensions'
import LRUCache from 'lru-cache'
const ONE_HOUR = 1000 * 60 * 60

const defaultCache = new LRUCache({ maxAge: ONE_HOUR })

export default function ({ $axios }) {
  const defaults = $axios.defaults
  // https://github.com/kuitos/axios-extensions
  defaults.adapter = cacheAdapterEnhancer(defaults.adapter, {
    enabledByDefault: true,
    defaultCache
  })
}
```

nuxt.config.js
```
plugins: [
    { src: '~/plugins/axios.js' },
]
```

So this should cache every route by axios for 1 hour.

/store/index.js
```
async nuxtServerInit ({ commit }, { $config }) {

        const wpheaders { 'x-requested-with' : 'XMLHttpRequest' }
        const bcheaders { 'X-Auth-Token': 'xxx', 'x-requested-with' : 'XMLHttpRequest' }

        return await axios.all([
            axios.get(`${$config.API_URL_BASE}/namespace/app`, { headers: wpheaders }),
            axios.get(`${$config.API_URL_BIG}/categories`, { headers: bcheaders }),
            axios.get(`${$config.API_URL_BIG}/products?page=1`, { headers: bcheaders }),
        ]).then(axios.spread((...responses) => {

            ...

            let productsArray = ( responses[2].data ) ? responses[2].data.data : [];

            console.log(
                'RATE LIMITS: ',
                responses[2]['headers']['x-rate-limit-requests-left'],
            );

            ...
            commit( 'SET_PRODUCTS', productsArray )

    }))
}
```
There's more calls here, but as you can see the rate limit quickly diminishes.

![Screenshot 2022-02-16 at 16 16 31](https://user-images.githubusercontent.com/5207147/154358201-84b23108-cf54-40b0-af4a-b6ca44726d7a.jpg)

Any help is appreciated as read the docs inside out and found nothing of use.

**Comment by danielroe** [comment]
I'd recommend [using a payload](https://nuxtjs.org/docs/2.x/configuration-glossary/configuration-generate/#speeding-up-dynamic-route-generation-with-payload) in this instance.

**Reply by CHEWX** [reply]
Didn't know discussions existed! Thanks.

`nuxtServerInit ({ commit }, { $config }) {`

and then using:

`$config.payload` returns undefined...?

I read https://github.com/nuxt/nuxt.js/issues/7037#issuecomment-595390230 here that it returns on second argument?

**Reply by danielroe** [reply]
It would be:
```js
nuxtServerInit ({ commit }, { $config, payload }) {
```

**Reply by CHEWX** [reply]
That just returns `null` or `{}`

**Reply by danielroe** [reply]
In dev, the payload will be empty; but it will be populated when generating. So you need to check and if it doesn't exist, fetch again...

**Reply by CHEWX** [reply]
I don't know how this applies to my setup though? Do I need to move `nuxtServerInit()` to `nuxt.config.js`? Sounds like a re-write.

**Reply by simonhrogers** [reply]
Hey Daniel thanks for this it’s very interesting! I understand and regularly use payloads in the context that’s described in your link; however I can’t see how a refactor would work in terms of `nuxtServerInit()` – sorry if painfully obvious – would you please be able to expand a little further? 

Can `nuxtServerInit()` be called from `nuxt.config.js`; or are you suggesting we make a regular API call there; pass this to every single page and then initiate the store someplace else the rest of the time... if I’m understanding correctly then this seems like an extremely big refactor of my project’s code for something you’d expect to be handled by `nuxtServerInit()`, so I guess I’m getting the wrong end of the stick somewhere. 

As an example, I’m building a website where several things need to be done on `nuxtServerInit()`:

- initiating a Shopify checkout
- fetching data once for a dynamic navigation component
- fetching a series of dynamic colour schemes which are used across the site
- fetching fallback SEO information which is used on all pages where none is defined

Shopify has rate-limiting, so fails on many pages being built, and the non-presence of navigation info when loading a dynamic route is causing me trouble too.

**Reply by simonhrogers** [reply]
Well sometimes things really are as simple as they sound... got this working – thank you! Will leave here for anyone else that finds and is silly like me... you just

In `store/index.js`:
```
export const actions = {
  async nuxtServerInit({ commit, dispatch, rootGetters }, { payload }) {
    if (payload) {
      commit('inventory/SET_INVENTORY', payload.inventory);
      commit('sanity/SET_NAVIGATION', payload.navigation);
      commit('sanity/SET_COLORSCHEMES', payload.colorSchemes);
      commit('sanity/SET_SEO', payload.seo);
    } else {
      await dispatch('inventory/INVENTORY_CALL');
      await dispatch('sanity/NAVIGATION_CALL');
      await dispatch('sanity/COLORSCHEMES_CALL');
      await dispatch('sanity/SEO_CALL');
    }
  }
}
```

In `nuxt.config.js`:

```
const dynamicRoutes = async() => {
  const { navigation } = await sanity.fetch(navigationRequest);
  const { colorSchemes } = await sanity.fetch(colorSchemesRequest);
  const seo = await sanity.fetch(seoRequest);
  const inventory = await fetchShopifyInventory(shopify);
  const payload = {
    navigation,
    colorSchemes,
    seo,
    inventory
  };
  const resForArticles = await sanity.fetch(articlePagesRequest);
  const routesForArticles = resForArticles.articles.map(article => {
    return {
      route: `/${ article.slug }`,
      payload: {
        item: article,
        ...payload
      }
    }
  });
 // ABRIDGED 
  
  ```

**Reply by CHEWX** [reply]
Thanks for following up @simonhrogers , will re-take a look at my setup, as have put some harsh timeouts for the time being to stop the flooding.

**Reply by simonhrogers** [reply]
No worries Tom :-) My setup is nothing flashy but in my case the Shopify API was rate-limiting and rather than build templates without it; this was actually interrupting the generation process for every subsequent dynamic page; causing the URLs to display little/no content – disaster...
